File: android/app/src/main/java/com/lightningkite/kwifttemplate/shared/views/LoginDemoVG.kt
package com.lightningkite.kwifttemplate.shared.views

import android.view.View
import com.lightningkite.kwift.actual.delay
import com.lightningkite.kwift.actual.weak
import com.lightningkite.kwift.observables.actual.bind
import com.lightningkite.kwift.observables.actual.bindLoading
import com.lightningkite.kwift.observables.actual.bindString
import com.lightningkite.kwift.observables.shared.MutableObservableProperty
import com.lightningkite.kwift.observables.shared.ObservableStack
import com.lightningkite.kwift.observables.shared.StandardObservableProperty
import com.lightningkite.kwift.views.actual.ViewDependency
import com.lightningkite.kwift.views.actual.onClick
import com.lightningkite.kwift.views.shared.*
import com.lightningkite.kwifttemplate.R
import com.lightningkite.kwifttemplate.layouts.LoginDemoXml

class LoginDemoVG(stack: ObservableStack<ViewGenerator>) : ViewGenerator() {
    val stack: ObservableStack<ViewGenerator>? by weak(stack)
    override val title: String get() = "Log In Demo"

    val form: Form = Form()
    val username: FormField<String> =
        form.field(R.string.username, "") { field -> field.required() }
    val password: FormField<String> =
        form.field(R.string.password, "") { field -> field.required() }
    val verifyPassword: FormField<String> =
        form.field(R.string.verify_password, "") { field -> field.required() ?: field.matches(password) }
    val agree: FormField<Boolean> =
        form.field(R.string.password, false) { field -> ViewStringResource(R.string.mustAgree).unless(field.value) }
    val loading: MutableObservableProperty<Boolean> = StandardObservableProperty(false)

    override fun generate(dependency: ViewDependency): View {
        val xml = LoginDemoXml()
        val view = xml.setup(dependency)

        xml.username.bindString(username.observable)
        xml.password.bindString(password.observable)
        xml.verifyPassword.bindString(verifyPassword.observable)
        xml.agree.bind(agree.observable)
        xml.submitLoading.bindLoading(loading)
        xml.submit.onClick {
            this.submit()
        }

        return view
    }

    private fun submit() {
        this.form.runOrDialog {
            println("Submit!")
            this.loading.value = true
            delay(1000) {
                this.loading.value = false
                this.stack?.push(ExampleContentVG())
            }
        }
    }
}

File: ios/Kwift Template/com/lightningkite/kwifttemplate/shared/views/LoginDemoVG.swift
//Package: com.lightningkite.kwifttemplate.shared.views
//Converted using Kwift2

import Foundation
import Kwift
import KwiftMaps



public class LoginDemoVG: ViewGenerator {
    
    
    public weak var stack: ObservableStack<ViewGenerator>? 
    override public var title: String {
        get {
            return "Log In Demo"
        }
    }
    public var form: Form
    public var username: FormField<String>
    public var password: FormField<String>
    public var verifyPassword: FormField<String>
    public var agree: FormField<Bool>
    public var loading: MutableObservableProperty<Bool>
    
    override public func generate(dependency: ViewDependency) -> View {
        var xml = LoginDemoXml()
        var view = xml.setup(dependency)
        xml.username.bindString(username.observable)
        xml.password.bindString(password.observable)
        xml.verifyPassword.bindString(verifyPassword.observable)
        xml.agree.bind(agree.observable)
        xml.submitLoading.bindLoading(loading)
        xml.submit.onClick{ () in 
            self.submit()
        }
        return view
    }
    override public func generate(_ dependency: ViewDependency) -> View {
        return generate(dependency: dependency)
    }
    
    private func submit() -> Void {
        self.form.runOrDialog{ () in 
            print("Submit!")
            self.loading.value = true
            delay(1000) { () in 
                self.loading.value = false
                self.stack?.push(ExampleContentVG())
            }
        }
    }
    
    public init(stack: ObservableStack<ViewGenerator>) {
        self.stack = stack
        let form: Form = Form()
        self.form = form
        let username: FormField<String> = form.field(ResourcesStrings.username, "") { (field) in 
            field.required()
        }
        self.username = username
        let password: FormField<String> = form.field(ResourcesStrings.password, "") { (field) in 
            field.required()
        }
        self.password = password
        let verifyPassword: FormField<String> = form.field(ResourcesStrings.verifyPassword, "") { (field) in 
            field.required() ?? field.matches(password)
        }
        self.verifyPassword = verifyPassword
        let agree: FormField<Bool> = form.field(ResourcesStrings.password, false) { (field) in 
            ViewStringResource(ResourcesStrings.mustAgree).unless(field.value)
        }
        self.agree = agree
        let loading: MutableObservableProperty<Bool> = StandardObservableProperty(false)
        self.loading = loading
        super.init()
    }
    convenience public init(_ stack: ObservableStack<ViewGenerator>) {
        self.init(stack: stack)
    }
}
 
